src/Algebra/Graph/Acyclic/Labelled/AdjacencyMap/Algorithm.hs

```diff
@@ -0,0 +1,47 @@
+module Algebra.Graph.Acyclic.Labelled.AdjacencyMap.Algorithm where
+
+import Algebra.Graph.Acyclic.Labelled.AdjacencyMap
+import qualified Algebra.Graph.AdjacencyMap.Algorithm as AM
+import Algebra.Graph.Label
+import qualified Algebra.Graph.Labelled.AdjacencyMap as LAM
+import Data.Map.Strict (Map, (!))
+import qualified Data.Map.Strict as Map
+import Data.Maybe (fromMaybe)
+import Data.List (foldl')
+
+-- TODO: Replace this function with 'skeleton' from Acyclic.Labelled to remove the use of fromMaybe
+-- TODO: Make 'topSort' more efficient
+-- TODO: Add examples and tests for 'topSort'
+topSort :: (Ord a) => AdjacencyMap e a -> [a]
+topSort = fromMaybe [] . AM.topSort . LAM.skeleton . fromAcyclic
+
+-- TODO: Improve documentation for 'fold'
+-- TODO: Add examples and tests for 'fold'
+-- | Compute the final state by using the initial state and
+-- traversing the entire graph in topological order.
+fold :: (Ord a) => (e -> a -> a -> s -> s) -> s ->  AdjacencyMap e a -> s
+fold f s am = foldl' f' s . concatMap unfold . topSort $ am
+  where
+    em = LAM.adjacencyMap . fromAcyclic $ am
+    unfold x = map (\(a, e) -> (e, x, a)) . Map.toList $ em ! x
```

Comment:
I don't think the order of processing edges is correct here. For example, for graph `1 * 2 * 3 * 4` the order will be:

`[(1,2), (1,3), (1,4), (2,3), (2,4), (3,4)]`

which processes the edge `1 -> 4` before the topologically earlier `2 -> 3`.

I think the order should be:

`[(1,2), (1,3), (2,3), (1,4), (2,4), (2,4)]`

Reply:
Ah, I think I interpreted it incorrectly. In my mind I had a notion of applying a function to all edges from (rather than to) a vertex in topological order. Will change it.

Reply:
@snowleopard I think we should change the signature of `fold` to something like `a -> [(e, a)] -> s -> s` as before. This is more general, as, with `e -> a -> a -> s -> s`, if there are no edges then we would fold over nothing.

Reply:
@adithyaov Yes, good idea. Indeed, we'd like to have a meaningful `fold` for `vertices xs`.

I suppose in `a -> [(e, a)] -> s -> s` we will call the function for every vertex `a` in the topological order, passing all outgoing edges `[(e, a)]` to it.

Reply:
I don't think it would matter much but I would prefer `[(e, a)]` to be all the incoming edges to the vertex. It gives a more clear picture and it's cleaner to visualize the flow.

For example, If you think of the acyclic graph as a dependency graph it is more comfortable to talk about a resource and its dependencies rather than a resource and all the resources that depend on it.

Reply:
Agreed, having a list of dependencies seems more natural. In this case I suggest swapping the order of arguments: `[(e, a)] -> a -> s -> s` -- in this way the first `a` refers to a source of an edge, and the second to its destination, just like in `edges` for example. Seems easier to remember.